### PR TITLE
Closes #6764: Refactor lcp beacon script

### LINKED ISSUE #6764: 3.17 - Refactor beacon script
# Description

The 3.17 feature introduce the Lazy Render Content (LRC). It will work the same way the LCP (3.16 introduced feature). Which means we will need a beacon script to run some logic, send an AJAX call to save data in the DB. However, as AJAX Call are costly in terms of performances due to the fact it loads WordPress for each call, we’ve decided to have only 1 beacon script that will manage both feature and regroup data to be sent in the same AJAX Call. 

In order to do this, we’ll have to refactor the current `lcp-beacon.js` script. 

The current `lcp-beacon.js` structure works with a unique class called `RocketLcpBeacon` . In the new structure of the beacon script, we will have 3 classes:

- `RocketBeacon`
- `RocketLcpBeacon`
- `RocketLrcBeacon`

The first one will be responsible for running everything, and will have the shared function across both features. As an example, both features requires to check the screen size, then the screen size function will be in `RocketBeacon`. 

While the two others (`RocketLcpBeacon` & `RocketLrcBeacon` ) will handle their own logic’s feature.

# Scope a solution

1. Rename the `lcp-beacon.js` to `wpr-beacon.js`.
    - Update the filename in the `buildLcpBeaconMin` method in the `src/js/gulp/tasks/js.js` file.
    - Update the filename in the `watch` method in the `src/js/gulp/tasks/js.js` file.
2. Create a new class `RocketBeacon` in the `wpr-beacon.js` file.
    - Implement the constructor for the `RocketBeacon` class. The constructor should take a `config` object as a parameter and initialize instance variables like `config`, `lcpBeacon`.
3. Implement the `init` method in the `RocketBeacon` class.
    - This method should check if the preconditions are valid using the `_isValidPreconditions` method.
    - If the preconditions are valid, it should check if the data has been generated before using the `_isGeneratedBefore` method.
    - If the data has not been generated before, it should create new instances of `RocketLcpBeacon, and run its process.
    - Finally, it should save the final result into the database using the `_saveFinalResultIntoDB` method.
4. Implement the `_isValidPreconditions`, `_isPageCached`, `_isGeneratedBefore`, and `_isNotValidScreensize` methods in the `RocketBeacon` class.
    - The `_isValidPreconditions` method should check the screen size and if the page is cached.
    - The `_isPageCached` method should check if the page is cached.
    - The `_isGeneratedBefore` method should make an AJAX call to check if there are any records for the current URL.
    - The `_isNotValidScreensize` method should check the screen size.
5. Implement the `_saveFinalResultIntoDB` method in the `RocketBeacon` class.
    - This method should create a new FormData object and append the necessary data to it.
    - It should then make an AJAX call to save the data into the database.
6. Implement the `run` method in the `RocketBeacon` class.
    - This method should create a new instance of the `RocketBeacon` class and call its `init` method.
7. Update the `RocketLcpBeacon` class to work with the new `RocketBeacon` class.
    - Update the constructor to take a `config` object as a parameter.
    - Update the `init` method to use the methods from the `RocketBeacon` class.
8. Thoroughly test the new `RocketBeacon`, `RocketLcpBeacon`, classes to ensure they work as expected.
9. Update all references to the old `lcp-beacon.js` file in your project to the new `wpr-beacon.js` file.
### Additional

You’ll find below an example on how the `wpr-beacon.js` will be built. This code hasn’t been tested and isn’t fully implemented:

Please note that `RocketLrcBeacon` shouldn't be implemented in this issue.

```jsx
class RocketBeacon {
  constructor(config) {
    this.config = config;
    this.lcpBeacon = null;
    this.lrcBeacon = null;
  }

  async init() {
    if (!this._isValidPreconditions()) {
      return;
    }

    const isGeneratedBefore = await this._isGeneratedBefore();

    if (!this._isNotValidScreensize()) {
      if (!isGeneratedBefore.lcp) {
        this.lcpBeacon = new RocketLcpBeacon(this.config);
        await this.lcpBeacon.run();
      }

			if (!isGeneratedBefore.lrc) {
	      this.lrcBeacon = new RocketLrcBeacon(this.config);
	      await this.lrcBeacon.run();
      }
    }

    this._saveFinalResultIntoDB();
  }

	async _isValidPreconditions() {
		// Check the screensize first because starting any logic.
		if ( this._isNotValidScreensize() ) {
			this._logMessage('Bailing out because screen size is not acceptable');
			return false;
		}

		if ( this._isPageCached() && await this._isGeneratedBefore() ) {
			this._logMessage('Bailing out because data is already available');
			return false;
		}

		return true;
	}
  
  _isPageCached() {
		const signature = document.documentElement.nextSibling && document.documentElement.nextSibling.data ? document.documentElement.nextSibling.data : '';

		return signature && signature.includes( 'Debug: cached' );
	}

	async _isGeneratedBefore() {
		// AJAX call to check if there are any records for the current URL.
		let data_check = new FormData();
		data_check.append('action', 'rocket_check_lcp');
		data_check.append('rocket_lcp_nonce', this.config.nonce);
		data_check.append('url', this.config.url);
		data_check.append('is_mobile', this.config.is_mobile);

		const beacon_data_response = await fetch(this.config.ajax_url, {
				method: "POST",
				credentials: 'same-origin',
				body: data_check
			})
			.then(data => data.json());
		return beacon_data_response.data;
	}

  _isNotValidScreensize() {
		// Check screen size
		const screenWidth = window.innerWidth || document.documentElement.clientWidth;
		const screenHeight= window.innerHeight || document.documentElement.clientHeight;

		const isNotValidForMobile = this.config.is_mobile &&
			( screenWidth > this.config.width_threshold || screenHeight > this.config.height_threshold );
		const isNotValidForDesktop = !this.config.is_mobile &&
			( screenWidth < this.config.width_threshold || screenHeight < this.config.height_threshold );

		return isNotValidForMobile || isNotValidForDesktop;
	}

  static _isIntersecting(rect) {
    // Check if any part of the image is within the viewport
    return (
      rect.bottom >= 0 &&
      rect.right >= 0 &&
      rect.top <= (window.innerHeight || document.documentElement.clientHeight) &&
      rect.left <= (window.innerWidth || document.documentElement.clientWidth)
    );
  }
  
  _saveFinalResultIntoDB() {
    const lcpResults = this.lcpBeacon ? this.lcpBeacon.getResults() : null;
    const lrcResults = this.lrcBeacon ? this.lrcBeacon.getResults() : null;

    const data = new FormData();
    data.append('action', 'rocket_lcp');
    data.append('rocket_lcp_nonce', this.config.nonce);
    data.append('url', this.config.url);
    data.append('is_mobile', this.config.is_mobile);
    data.append('images', JSON.stringify(this.performanceImages));
    data.append('status', this._getFinalStatus());
    data.append('lcpResults', JSON.stringify(lcpResults));
    data.append('lrcResults', JSON.stringify(lrcResults));
    // rest of the AJAX call code...
	}
}

class RocketLcpBeacon {
  constructor(config) {
    this.config = config;
  }

  async run() {
	  try {
			const above_the_fold_images = this._generateLcpCandidates( Infinity );
			if ( above_the_fold_images ) {
				this._initWithFirstElementWithInfo( above_the_fold_images );
				this._fillATFWithoutDuplications( above_the_fold_images );
			}
		} catch ( err ) {
			this.errorCode = 'script_error';
			this._logMessage( 'Script Error: ' + err );
		}
  }
  
  _generateLcpCandidates( count ) {
	  // Contains the logic behind lcp feature
  }
  
   _getResults() {
	  // return the results from this feature
  }
}

// not getting added within the refactor
class RocketLrcBeacon {
  constructor(config) {
    this.config = config;
  }

  async run() {
    // LRC-specific logic
    // This might include logic for identifying LRC elements, storing their information, etc.
    // You can use RocketBeacon._isIntersecting(rect) for any common functionality needed.
  }
  
  _getResults() {
	  // return the results from this feature
  }
}

RocketBeacon.init();
```

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing included e2e tests.

### PR DESCRIPTION
**Pull request by Miraeld**
Change the structure of the beacon script to include a new class `RocketBeacon` which will contains more generic functions that could be shared with other features.

# Description

Fixes #6764

### User documentation

This shouldn't impact users.

### Technical documentation

The beacon script has been renamed to `wpr-beacon.js` & has been refactored. 
This refactor includes the addition of a new class named `RocketBeacon` which contains generic functions used later on by `RocketLcpBeacon` class. 
It will allow us to add more feature easily to avoid duplication. 
In order to test this PR, https://github.com/wp-media/wp-rocket-e2e/issues/124 needs to be merged into e2e. 

## Type of change
- [x] Enhancement (non-breaking change which improves an existing functionality).

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Documentation

- [x] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [x] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [x] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.
- [x] I listed the introduced external dependencies explicitely on the PR.
- [x] I validated the repo-specific guidelines from CONTRIBUTING.md.

## Observability
- [x] I handled errors when needed.
- [x] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [x] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [x] I explicitely mentioned performance risks in the PR.
- [x] I explicitely mentioned security risks in the PR.

### CODE REVIEW

File: assets/js/wpr-beacon.js [hunk 1]
```diff
@@ -49,52 +43,161 @@ class RocketLcpBeacon {
...
 		const signature = document.documentElement.nextSibling && document.documentElement.nextSibling.data ? document.documentElement.nextSibling.data : '';
-
-		return signature && signature.includes( 'Debug: cached' );
+		return signature && signature.includes('Debug: cached');
 	}
 
 	async _isGeneratedBefore() {
-		// AJAX call to check if there are any records for the current URL.
 		let data_check = new FormData();
 		data_check.append('action', 'rocket_check_lcp');
```
Comment (jeawhanlee) [inline assets/js/wpr-beacon.js]: Should `rocket_check_lcp` not be changing to something more generic since its single ajax request?

**Review by remyperona — APPROVED** [review]

File: assets/js/wpr-beacon.js [hunk 2]
```diff
@@ -49,52 +43,161 @@ class RocketLcpBeacon {
...
+			rect.top <= (window.innerHeight || document.documentElement.clientHeight) &&
+			rect.left <= (window.innerWidth || document.documentElement.clientWidth)
+		);
+	}
+
+	_saveFinalResultIntoDB() {
+		const lcpResults = this.lcpBeacon ? this.lcpBeacon.getResults() : null;
+
+		const data = new FormData();
+		data.append('action', 'rocket_lcp');
```
Comment (jeawhanlee) [inline assets/js/wpr-beacon.js]: This also, `rocket_lcp` should it not be changed to something more generic?

**Comment by remyperona** [comment]
Should this PR be done against `develop`? Shouldn't it be against the refactor branch?

**Comment by Miraeld** [comment]
@jeawhanlee I've updated action names and everything related to that. 

@Tabrisrp I've changed the based branch.

File: inc/Engine/Media/AboveTheFold/Frontend/Controller.php [hunk 3]
```diff
@@ -412,15 +412,15 @@ public function inject_beacon( $html, $url, $is_mobile ): string {
 		 *
 		 * @param int $delay The delay in milliseconds. Default is 500.
 		 */
-		$delay = apply_filters( 'rocket_lcp_delay', $default_delay );
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/Frontend/Controller.php]: if we change the filter name, we should add an apply_filters_deprecated call too for backward compatibility

**Review by remyperona — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -32,20 +32,21 @@ To be able to validate your license and use the plugin, you will also have to ma
...
+| `gulp build:saas:min`           |                     Builds Full admin CSS, the minified version (wpr-admin.min.css)                     |
+| `gulp build:sass:all`           | Builds all admin CSS files (wpr-admin.css, wpr-admin.min.css, wpr-admin-rtl.css, wpr-admin-rtl.min.css) |
+| `gulp sass:watch`               |           Watches all admin CSS files mentioned above and builds them again with any change.            |
+| **JS Tasks**                    |                                                                                                         |
+| `gulp build:js:app:unmin`       |                     Builds admin app js file, the unminified version (wpr-admin.js)                     |
+| `gulp build:js:app:min`         |                    Builds admin app js file, the minified version (wpr-admin.min.js)                    |
+| `gulp build:js:lazyloadcss:min` |                 Builds lazyload CSS js file, the minified version (lazyload-css.min.js)                 |
+| `gulp build:js:lcp`             |      Builds lcp beacon script, the minified version (lcp-beacon.min.js, source file, and map file)      |
```
Comment (remyperona) [inline README.md]: I think this line should be removed?

**Review by remyperona — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (wordpressfan) [reply README.md]: Good catch, Done

**Review by remyperona — APPROVED** [review]

**Pull request merged**